docs/RFCS/leader_evaluated_raft.md

```diff
@@ -0,0 +1,425 @@
+    # * remove some metadata                            [move pre-raft]
+    #   careful: this is a DeleteRange; could benefit from a ClearRange
+    #   Raft effect (all inline values; no versioning needed)
+    # * clear tsCache (not really necessary)
+    optional Merge ...:
+    # Carries out the following:
+    # * `(*Replica).setDesc`
+    # * maybe add to GC queue                           [leader only]
+    optional ChangeReplicas ...;
+    # Carries out the following:
+    # * maybe gossips system config                     [leader only]
+    optional ModifiedSpan ...;
+    # Carries out the following:
+    # * sanity checks (in particular FindReplica)
+    # * update in-memory lease
+    # * on Raft(!) leader: maybe try to elect new lease holder
+    # * on (new) lease holder: update tsCache low water mark
+    #   (might be easier to do this unconditionally on all nodes)
+    # * on (new) lease holder: maybe gossip system configs
+    optional LeaderLease ...;
+    # Those below may not be required, but something to think about
+    optional GCRange ...;    # allow optimized GC (some logic below of Raft?)
+    optional ClearRange ...; # allow erasing key range without huge proposal
+  }
+}
+```
+
+Since this simple structure is easy to reason about, it seems adequate to
+strive for the minimum amount of interaction between the various effects
+(though the data required for the triggers may benefit from more interaction
+between them in some cases)
+
+### Computing the (ordered) write set
+
+Before a leader proposes a command, it has the job of translating a
+`BatchRequest` into a `RaftCmd`. Note that we assume that writes wait for
+both reads and writes in the command queue (so that at the point at which
+the `RaftCmd` is constructed, all prior mutations have been carried out and
+the timestamp cache updated).
+
+This largely takes over the work which was done in
+
+```
+applyRaftCommandInBatch -> executeWriteBatch -> executeBatch
+                        -> (ReplicaCommands) -> (MVCC)
+```
+
+This is hopefully largely mechanical, with the bulk of work in the MVCC layer,
+outlined in the next section.
+
+The work to be carried out in `ReplicaCommands` is mostly isolating the various
+side effects (most prominently the commit and leader lease triggers). In the
+process, it should be possible to remove these methods from `(*Replica)`
+(though that is not a stated goal of this RFC).
+
+### MVCC & Engine Layer
+
+Operations in the MVCC layer (`MVCC{Get,Put,Scan,DeleteRange,...}`) currently
+return some variation of `([]roachpb.KeyValue, []roachpb.Intent, error)` and
+operate on the underlying `engine.Engine` (which in practice is a RocksDB
+batch) with heavy use of iterators.
+
+With the new code, we want mostly the same, but with two important changes:
+
+* we do not actually intent to commit the batch (though we could do this on the
+  proposing node to avoid overhead)
+* we need to construct the `Writes` field, which (in some way or
```

Comment:
I don't think you'll need any of the changes proposed in this MVCC section (as you point out in the section proposing the use of RocksDB's `WriteBatch` format).